minor-modes.lisp

```diff
@@ -0,0 +1,1083 @@
+                               (attempt-binding key rest bind seq)))))
+               (if (not (or (symbolp bind-to)
+                            (stringp bind-to)
+                            (functionp bind-to)))
+                   (restart-case (error "Invalid binding ~A" bind-to)
+                     (bind-anyway ()
+                       :report "Bind the key binding regardless"
+                       (traverse-and-bind in-seq))
+                     (skip-binding ()
+                       :report "skip this binding"
+                       nil))
+                   (traverse-and-bind in-seq))))))
+      (cond ((null keymap)
+             topmap)
+            ((or (symbolp keymap)
+                 (stumpwm::kmap-p keymap))
+             keymap)
+            ((listp keymap)
+             (restart-case (mapc (lambda (keys)
+                                   (create-keymap-binding keys)
+                                   (setf curmap topmap))
+                                 keymap)
+               (abort-bindings ()
+                 :report "Return the keymap without binding further keys"
+                 topmap)
+               (abort-bindings* ()
+                 :report "Return an empty keymap"
+                 (stumpwm:make-sparse-keymap)))
+             topmap)
+            (t (restart-case
+                   (error "Function MAKE-MINOR-MODE-KEYMAP cant understand ~A"
+                          keymap)
+                 (use-empty-keymap ()
+                   :report "Use an empty keymap"
+                   (stumpwm:make-sparse-keymap))))))))
+
+(defun make-minor-mode-keymap (spec)
+  (generate-keymap spec))
+
+(defun make-minor-mode-top-map (top-map-spec root-map-spec)
+  "Create a top map for a minor mode based upon its TOP-MAP-SPEC and
+ROOT-MAP-SPEC."
+  (let ((top-map nil)
+        (root-map (if root-map-spec
+                      (make-minor-mode-keymap root-map-spec)
+                      (make-sparse-keymap))))
+    (fill-keymap top-map *escape-key* root-map)
+    (generate-keymap top-map-spec top-map)))
+
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-special-variable-name (mode name)
+    (intern (format nil "*~A-~A*" mode name)))
+  
+  (defun parse-minor-mode-options (options)
+    (let ((valid-options
+            '((:interactive            1)
+              (:scope                  1)
+              (:global                 1)
+              (:lighter-make-clickable 1)
+              (:lighter                1)
+              (:expose-keymaps         1)
+              (:root-map               1)
+              (:top-map                1)
+              (:enable-when            t)
+              (:make-hooks             1)
+              (:default-initargs       t)
+              (:define-command-definer 1)))
+          (all-vals '())
+          (other-opts '()))
+      (flet ((collect-values (option)
+               (let ((argcount (cadr (assoc (car option) valid-options))))
```

Comment:
I think using destructuring-bind here would be more clear and avoid the need to use all the cons manipulation operations.

Reply:
I'm not sure I quite follow - do you mean to use destructuring-bind on the options we receive? I'm not sure that would work, as the options are received in the form `((:o1 a b c) (:o2 d e f))`, and I dont think destructuring bind can handle nested lists in an arbitrary order like that. I may not be understanding you correctly.

Reply:
I meant instead of stuff like (cadr option), (car option) etc. Using d-bind we can bind the key and value of the current option at the start of the function, instead of the let.

Also iiuc we don't need to use assoc as the count would be the cdr of the current option being processed right?

Reply:
Ah I was misunderstanding you. Yes d-bind would work there and I'll add it in. I do think we need to use assoc as this is constructing a list to (eventually) pass to d-bind in the main `define-minor-mode` macro. I want to keep things open ended with regards to adding options to this macro, and if we ditched assoc we would have no way to discriminate between a future 2 argument option and an option with an arbitrary number of arguments. Since this is all happening at macroexpansion time it doesnt have all that much impact on performance.

Reply:
> I do think we need to use assoc as this is constructing a list to (eventually) pass to d-bind in the main define-minor-mode macro. I want to keep things open ended with regards to adding options to this macro, and if we ditched assoc we would have no way to discriminate between a future 2 argument option and an option with an arbitrary number of arguments.

I may be missing something, but isn't the open ended design achieved by the encoding the semantics in data (the alist)? That description is then processed one at the time. But if we change the valid-options alist we still need to recompile for it take effect. Wether we use the assoc or not.

Reply:
I think I understand what you're getting at (correct me if im wrong); Since we want to return the `cadr` of 1 arg options and the `cdr` of n arg options we can change options to be passed as `(:foo . bar)` for 1 arg options and `(:baz qux)` for n arg options, and then just take the cdr of all options instead of looking them up in an alist. Is that what you're talking about? 

I am somewhat attached to using an alist - this parsing was heavily inspired by the function `parse-define-application-frame-options` from the [mcclim codebase](https://github.com/McCLIM/McCLIM/blob/master/Core/clim-core/frames/define-application-frame.lisp#L138), and having the ability to easily add additional information without restructuring the core of how we parse options is something I'd like to hold on to. For example if in the future we wish to add the option `:define-hooks` which defines only specific hooks and conflicts with `:make-hooks`, we could add conflict checking by modifying the alist and `collect-values`.

Reply:
>  and then just take the cdr of all options instead of looking them up in an alist. Is that what you're talking about?

yes, this is what I mean.

> I am somewhat attached to using an alist

I haven't suggested otherwise, I think using an alist is a great idea. Pushing semantics to a declarative description is great in my book. My issue is what you mentioned above, that we don't need to perform a lookup as we can get all the information we want by d-binding the description.

> this parsing was heavily inspired by the function parse-define-application-frame-options from the [mcclim codebase](https://github.com/McCLIM/McCLIM/blob/master/Core/clim-core/frames/define-application-frame.lisp?rgh-link-date=2022-06-23T08%3A05%3A44Z#L138),

interesting the linked code uses an alist of (almost) plists, which is why it uses getf to access information inside the description. In our case the value of the mapping is a single value so no need to.

What do you think?

Reply:
I'm sorry, I think I must be misunderstanding you, either in part or in full. I thought you were referring to dbinding the `option` variable within the local function `collect-values`. Do you think you could give me a small code example of what you intend with dbind? 

To my understanding this branch of the conversation came from you saying we dont need to use assoc, but if we dont use assoc then the alist will never get used, so any information contained there would be best put in a comment or docstring. But then you say that using the alist is a good idea, so I'm fairly certain I have misunderstood something along the way. 

Sorry this communication is proving so difficult; if its useful we could "reset" the thread and just start from the beginning again.

As a side note, I dont think taking options as described in my prior comment would be the smoothest experience for people defining minor modes, eg i prefer
```lisp
(define-minor-mode mymode ()
  (:scope :frame)
  (:enable-when (mode obj)
   (something obj)))
```
as opposed to 
```lisp
(define-minor-mode mymode () 
  (:scope . :frame)
  (:enable-when (mode obj)
   (something obj)))
```